### PR TITLE
chat attachment support (cli, service)

### PR DESCRIPTION
**Pull request by patrickxb**
Sorry this is enormous.

This is the CLI and service side of chat attachments.

Attachments are uploaded/downloaded directly to/from S3 via the service daemon.  The payload header is signed by gregord to ensure clients have no access to S3 secret key.  Attachments > 5MB are uploaded in parts using the S3 multi API.  There is a 5TB limit to chat attachments.

Progress is reported via the ChatUI.

Currently there is no encryption/decryption as that is being handled by @oconnor663 in a separate ticket.

There is support for upload/download in `keybase chat api`, `keybase chat upload`, `keybase chat download`.

A gregor PR accompanies this.

r? @mmaxim @maxtaco 
cc @oconnor663

### CODE REVIEW

File: go/chat/backoff.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
+package chat
+
+// from here:
+// https://blog.gopheracademy.com/advent-2014/backoff/
+
+import (
+	"math/rand"
+	"time"
+)
+
```
Comment (mmaxim) [inline go/chat/backoff.go]: Maybe this should go in libkb? Might get use out of it in other places.

File: go/service/chat_local.go [hunk 2]
```diff
@@ -793,3 +933,42 @@ func (h *chatLocalHandler) assertLoggedIn(ctx context.Context) error {
...
+// Sign implements github.com/keybase/go/chat/s3.Signer interface.
+func (h *chatLocalHandler) Sign(payload []byte) ([]byte, error) {
+	arg := chat1.S3SignArg{
+		Payload: payload,
+		Version: 1,
+	}
+	return h.remoteClient().S3Sign(context.Background(), arg)
+}
+
+func (h *chatLocalHandler) uploadAsset(ctx context.Context, sessionID int, params chat1.S3Params, local chat1.LocalSource, progress chat.ProgressReporter) (chat1.Asset, error) {
```
Comment (patrickxb) [inline go/service/chat_local.go]: I added a ticket for that.  CORE-3959.  The Multi api supports it, but I didn't write the code for resuming (yet).

Comment (mmaxim) [inline go/service/chat_local.go]: Sounds good!

File: go/service/chat_local.go [hunk 3]
```diff
@@ -745,6 +747,144 @@ func (h *chatLocalHandler) PostLocal(ctx context.Context, arg chat1.PostLocalArg
...
+	if err != nil {
+		return chat1.PostLocalRes{}, err
+	}
+
+	// upload attachment and (optional preview) concurrently
+	var wg sync.WaitGroup
+	var objectErr, previewErr error
+	var object chat1.Asset
+	var preview *chat1.Asset
+	wg.Add(1)
```
Comment (patrickxb) [inline go/service/chat_local.go]: perhaps ErrGroup better to handle cancellation.  I'll change.

File: go/chat/attachment.go [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+	}
+
+	type job struct {
+		block []byte
+		index int
+	}
+	eg, ctx := errgroup.WithContext(ctx)
+	blockCh := make(chan job)
+	retCh := make(chan s3.Part)
+	eg.Go(func() error {
```
Comment (patrickxb) [inline go/chat/attachment.go]: Yeah, and if you watch the logs it really does help as nothing is held up by a single part upload being slow.  You'll see sometimes part 16 finishing before part 4, for example.

**Comment by patrickxb** [comment]
As far as the download progress, it does exactly 100.  I was envisioning a gui progress bar.  I'll make it less annoying in CLI though.

File: go/chat/attachment.go (hunk 4 above)
Comment (mmaxim) [inline go/chat/attachment.go]: Yeah very cool!

**Comment by mmaxim** [comment]
If you merge master back in you will get rid of that CI failure too.

**Comment by patrickxb** [comment]
@mmaxim all comments addressed, PTAL.

**Pull request merged**